controller/src/main/java/io/pravega/controller/store/index/PravegaTablesHostIndex.java

```diff
@@ -0,0 +1,134 @@
+    }
+
+    private CompletableFuture<Version> handleTableNotExist(String hostId, String entity, byte[] entityData, String hostEntityTable) {
+        return storeHelper.createTable(NameUtils.INTERNAL_SCOPE_NAME, hostsTable)
+                          .thenCompose(x -> storeHelper.addNewEntryIfAbsent(NameUtils.INTERNAL_SCOPE_NAME, hostsTable, hostId, new byte[0]))
+                          .thenCompose(x -> storeHelper.createTable(NameUtils.INTERNAL_SCOPE_NAME, hostEntityTable))
+                          .thenCompose(x -> storeHelper.addNewEntryIfAbsent(NameUtils.INTERNAL_SCOPE_NAME, hostEntityTable, entity, entityData));
+    }
+
+    private String getHostEntityTableName(String hostId) {
+        return String.format(HOST_TABLE_FORMAT, indexName, hostId);
+    }
+
+    @Override
+    public CompletableFuture<byte[]> getEntityData(String hostId, String entity) {
+        Preconditions.checkNotNull(hostId);
+        Preconditions.checkNotNull(entity);
+        String table = getHostEntityTableName(hostId);
+        return Futures.exceptionallyExpecting(storeHelper.getEntry(NameUtils.INTERNAL_SCOPE_NAME, table, entity)
+                .thenApply(Data::getData), DATA_NOT_FOUND_PREDICATE, null);
+    }
+
+    @Override
+    public CompletableFuture<Void> removeEntity(final String hostId, final String entity, final boolean deleteEmptyHost) {
+        Preconditions.checkNotNull(hostId);
+        Preconditions.checkNotNull(entity);
+        String table = getHostEntityTableName(hostId);
+        return Futures.exceptionallyExpecting(storeHelper.removeEntry(NameUtils.INTERNAL_SCOPE_NAME, table, entity),
+                DATA_NOT_FOUND_PREDICATE, null);
+    }
+
+    @Override
+    public CompletableFuture<Void> removeHost(final String hostId) {
+        Preconditions.checkNotNull(hostId);
+        String table = getHostEntityTableName(hostId);
+        return Futures.exceptionallyExpecting(Futures.exceptionallyExpecting(
+                storeHelper.deleteTable(NameUtils.INTERNAL_SCOPE_NAME, table, true).thenApply(v -> true), 
+                DATA_NOT_EMPTY_PREDICATE, false), DATA_NOT_FOUND_PREDICATE, true)
+                .thenCompose(deleted -> {
+                    if (deleted) {
+                        return Futures.exceptionallyExpecting(storeHelper.removeEntry(NameUtils.INTERNAL_SCOPE_NAME, hostsTable, hostId), DATA_NOT_FOUND_PREDICATE, null);
+                    } else {
+                        return CompletableFuture.completedFuture(null);
+                    }
+                });
+    }
+
+    @Override
+    public CompletableFuture<List<String>> getEntities(final String hostId) {
+        Preconditions.checkNotNull(hostId);
+        String table = getHostEntityTableName(hostId);
+        List<String> entries = new ArrayList<>();
+        return Futures.exceptionallyExpecting(storeHelper.getAllKeys(NameUtils.INTERNAL_SCOPE_NAME, table)
+                .collectRemaining(entries::add)
```

Comment:
@tkaitchuck 
> This involves multiple threads and needs to be made thread safe. I think a better solution would be to implement a 'collect' method on async iterator.
we are actually looping over the AsyncIterator elements one at a time and updating the record in the list. So only one thread is accessing the list at a given time. 
so we should be ok as far as synchronization is concerned here.